Issues running finetuned versions of supported models

I've been testing running various finetuned versions of supported models on GKE. However, it gets stuck on ` Using the Hugging Face API to retrieve tokenizer config` 

This are the full logs 

```
2024-07-02T18:54:46.894948Z  INFO text_generation_launcher: Args { model_id: "ICTNLP/Llama-2-7b-chat-TruthX", revision: None, validation_workers: 2, sharded: None, num_shard: None, quantize: None, speculate: None, dtype: None, trust_remote_code: false, max_concurrent_requests: 4, max_best_of: 2, max_stop_sequences: 4, max_top_n_tokens: 5, max_input_tokens: None, max_input_length: Some(32), max_total_tokens: Some(64), waiting_served_ratio: 1.2, max_batch_prefill_tokens: None, max_batch_total_tokens: None, max_waiting_tokens: 20, max_batch_size: Some(1), cuda_graphs: None, hostname: "tgi-tpu-84bdb4d847-fdw7j", port: 80, shard_uds_path: "/tmp/text-generation-server", master_addr: "localhost", master_port: 29500, huggingface_hub_cache: Some("/data"), weights_cache_override: None, disable_custom_kernels: false, cuda_memory_fraction: 1.0, rope_scaling: None, rope_factor: None, json_output: false, otlp_endpoint: None, cors_allow_origin: [], watermark_gamma: None, watermark_delta: None, ngrok: false, ngrok_authtoken: None, ngrok_edge: None, tokenizer_config_path: None, disable_grammar_support: false, env: false, max_client_batch_size: 4 }
2024-07-02T18:54:46.895049Z  INFO hf_hub: Token file not found "/root/.cache/huggingface/token"    
2024-07-02T18:54:47.190379Z  INFO text_generation_launcher: Default `max_batch_prefill_tokens` to 32
2024-07-02T18:54:47.190398Z  INFO text_generation_launcher: Using default cuda graphs [1, 2, 4, 8, 16, 32]
2024-07-02T18:54:47.190512Z  INFO download: text_generation_launcher: Starting download process.
2024-07-02T18:54:47.270672Z  WARN text_generation_launcher: 'extension' argument is not supported and will be ignored.

2024-07-02T18:54:51.295843Z  INFO download: text_generation_launcher: Successfully downloaded weights.
2024-07-02T18:54:51.296183Z  INFO shard-manager: text_generation_launcher: Starting shard rank=0
2024-07-02T18:54:53.940271Z  INFO text_generation_launcher: Server started at unix:///tmp/text-generation-server-0

2024-07-02T18:54:54.000595Z  INFO shard-manager: text_generation_launcher: Shard ready in 2.703388509s rank=0
2024-07-02T18:54:54.098804Z  INFO text_generation_launcher: Starting Webserver
2024-07-02T18:54:54.121130Z  INFO text_generation_router: router/src/main.rs:185: Using the Hugging Face API
2024-07-02T18:54:54.121182Z  INFO hf_hub: /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/hf-hub-0.3.2/src/lib.rs:55: Token file not found "/root/.cache/huggingface/token"    
2024-07-02T18:54:54.436475Z  INFO text_generation_router: router/src/main.rs:471: Serving revision 2d186e966af6eaa237495a39433a6f6d7de3ad9e of model ICTNLP/Llama-2-7b-chat-TruthX
2024-07-02T18:54:54.436496Z  INFO text_generation_router: router/src/main.rs:253: Using config Some(Llama)
2024-07-02T18:54:54.436501Z  INFO text_generation_router: router/src/main.rs:265: Using the Hugging Face API to retrieve tokenizer config
```

I get this issue in the following models: [RLHFlow/ArmoRM-Llama3-8B-v0.1](https://huggingface.co/RLHFlow/ArmoRM-Llama3-8B-v0.1), [Trendyol/Trendyol-LLM-7b-base-v0.1](https://huggingface.co/Trendyol/Trendyol-LLM-7b-base-v0.1), [ICTNLP/Llama-2-7b-chat-TruthX](https://huggingface.co/ICTNLP/Llama-2-7b-chat-TruthX). 

However, I've also been able to successfully run the following models: [UCLA-AGI/Llama-3-Instruct-8B-SPPO-Iter3](https://huggingface.co/UCLA-AGI/Llama-3-Instruct-8B-SPPO-Iter3), [georgesung/llama2_7b_chat_uncensored](https://huggingface.co/georgesung/llama2_7b_chat_uncensored)

I was wondering if there were any requirements in terms of files that were required to run a finetuned model? Or any help debugging the issue

This is the yaml that I used to deploy 

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: tgi-tpu
spec:
  replicas: 1
  selector:
    matchLabels:
      app: tgi-tpu
  template:
    metadata:
      labels:
        app: tgi-tpu
    spec:
      nodeSelector:
        cloud.google.com/gke-tpu-topology: 2x4
        cloud.google.com/gke-tpu-accelerator: tpu-v5-lite-podslice
      volumes:
        - name: data-volume
          emptyDir: {}
      containers:
      - name: tgi-tpu
        image: {optium_tpu_image}
        args:
        - --model-id=ICTNLP/Llama-2-7b-chat-TruthX
        - --max-concurrent-requests=4
        - --max-input-length=32
        - --max-total-tokens=64
        - --max-batch-size=1
        securityContext:
            privileged: true
        env:
          - name: HF_TOKEN
            value: {my_hf_token}
        ports:
        - containerPort: 80
        volumeMounts:
            - name: data-volume
              mountPath: /data
        resources:
          limits:
            google.com/tpu: 8
---
apiVersion: v1
kind: Service
metadata:
  name: service
spec:
  selector:
    app: tgi-tpu
  ports:
    - name: http
      protocol: TCP
      port: 8080  
      targetPort: 80  

```

Looked into this with @Edwinhr716 . Optimum tpu is failing to load these models for various reasons:

[RLHFlow/ArmoRM-Llama3-8B-v0.1](https://huggingface.co/RLHFlow/ArmoRM-Llama3-8B-v0.1)
```
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 818, in _mp_fn
    generator = TpuGeneratorSingleThread.from_pretrained(model_path, revision, max_batch_size, max_sequence_length)
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 776, in from_pretrained
    model = AutoModelForCausalLM.from_pretrained(
  File "/opt/optimum-tpu/optimum/tpu/modeling.py", line 64, in from_pretrained
    model = cls.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py", line 3626, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
  File "/opt/optimum-tpu/optimum/tpu/modeling_llama.py", line 1182, in __init__
    self.lm_head = ColumnParallelLinear.create(
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 605, in create
    return ColumnParallelLinear(
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 496, in __init__
    self.output_size_per_partition = divide_and_check_no_remainder(out_features, self.world_size)
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 218, in divide_and_check_no_remainder
    ensure_divisibility(numerator, denominator)
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 212, in ensure_divisibility
    assert numerator % denominator == 0, "{} is not divisible by {}".format(numerator, denominator)
AssertionError: 128257 is not divisible by 4
```

[Trendyol/Trendyol-LLM-7b-base-v0.1](https://huggingface.co/Trendyol/Trendyol-LLM-7b-base-v0.1)

```
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 71, in _thread_fn
    return fn()
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 187, in __call__
    self.fn(runtime.global_ordinal(), *self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 818, in _mp_fn
    generator = TpuGeneratorSingleThread.from_pretrained(model_path, revision, max_batch_size, max_sequence_length)
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 776, in from_pretrained
    model = AutoModelForCausalLM.from_pretrained(
  File "/opt/optimum-tpu/optimum/tpu/modeling.py", line 64, in from_pretrained
    model = cls.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py", line 3626, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
  File "/opt/optimum-tpu/optimum/tpu/modeling_llama.py", line 1182, in __init__
    self.lm_head = ColumnParallelLinear.create(
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 605, in create
    return ColumnParallelLinear(
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 496, in __init__
    self.output_size_per_partition = divide_and_check_no_remainder(out_features, self.world_size)
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 218, in divide_and_check_no_remainder
    ensure_divisibility(numerator, denominator)
  File "/opt/optimum-tpu/optimum/tpu/xla_model_parallel.py", line 212, in ensure_divisibility
    assert numerator % denominator == 0, "{} is not divisible by {}".format(numerator, denominator)
AssertionError: 44222 is not divisible by 4
```


 [ICTNLP/Llama-2-7b-chat-TruthX](https://huggingface.co/ICTNLP/Llama-2-7b-chat-TruthX)

```
File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 71, in _thread_fn
    return fn()
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 187, in __call__
    self.fn(runtime.global_ordinal(), *self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 818, in _mp_fn
    generator = TpuGeneratorSingleThread.from_pretrained(model_path, revision, max_batch_size, max_sequence_length)
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/generator.py", line 776, in from_pretrained
    model = AutoModelForCausalLM.from_pretrained(
  File "/opt/optimum-tpu/optimum/tpu/modeling.py", line 64, in from_pretrained
    model = cls.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py", line 3305, in from_pretrained
    raise EnvironmentError(
OSError: Error no file named pytorch_model.bin, model.safetensors, tf_model.h5, model.ckpt.index or flax_model.msgpack found in directory /data/models--ICTNLP--Llama-2-7b-chat-TruthX/snapshots/2d186e966af6eaa237495a39433a6f6d7de3ad9e.
```

For some reason the stdout or stderr logs are not getting redirected back to the console.

--- a/Makefile
+++ b/Makefile
@@ -66,6 +66,7 @@ style_check:
 	ruff check .
 
 style:
+	ruff check . --fix
 
 # Utilities to release to PyPi
 build_dist_install_tools:

--- a/text-generation-inference/server/text_generation_server/jetstream_pt_support/engine_loader.py
+++ b/text-generation-inference/server/text_generation_server/jetstream_pt_support/engine_loader.py
@@ -6,6 +6,7 @@ import time
 from typing import TYPE_CHECKING, Any
 
 import jax
+import jax.numpy as jnp
 import torch
 from jetstream_pt import fetch_models, quantize_model, torchjax
 from jetstream_pt.engine import PyTorchEngine
@@ -140,16 +141,40 @@ def instantiate_model_from_repo_id(
     return model
 
 
+def _get_needed_padding(value: int, multiple: int) -> int:
+    return (multiple - value % multiple) % multiple
+
+def _pad_array_up_to(v: jnp.ndarray, axis: int, multiple: int) -> jnp.ndarray:
+    a = [(0, 0) for _ in range(len(v.shape))]
+    a[axis] = (0, _get_needed_padding(v.shape[axis], multiple))
+    return jnp.pad(v, a)
+
+def pad_to_shard(env, val, axis: int):
+    # if axis is -1, then no sharding is done, everything is replicated
+    if axis == -1 or axis is None:
+        return val
+    sharding = env.sharding_by_axis(axis)
+    axis_name = sharding.spec[axis]
+    size_to_pad = env.mesh.shape[axis_name]
+    padded_val = _pad_array_up_to(val, axis, size_to_pad)
+    # Note that, even if the the value is padded, the model will only use the part that is needed.
+    if val.shape != padded_val.shape:
+        logger.debug(f"Sharding resulting in padding weights from {val.shape} to {padded_val.shape}")
+    return padded_val
+
+
 def shard_weights(env, weights, weight_shardings):
     """Shard weights according to weight_shardings"""
     for k, v in weight_shardings.items():
         logger.debug(f"SHARDING {k} {v}")
     sharded = {}
     for key, val in weights.items():
-        sharding = env.sharding_by_axis(weight_shardings.get(key, -1))
+        axis = weight_shardings.get(key, -1)
+        sharding = env.sharding_by_axis(axis)
         with jax.default_device(jax.devices("cpu")[0]):
             # Note we clone to avoid a core-dump that might happen otherwise when calling device_put
             arr = torch_xla2.tensor.t2j(val.clone())
+            arr = pad_to_shard(env, arr, axis)
         arr = jax.device_put(arr, sharding)
         sharded[key] = torchjax.to_torch(arr)
     return sharded

--- a/text-generation-inference/tests/test_decode_jetstream.py
+++ b/text-generation-inference/tests/test_decode_jetstream.py
@@ -59,8 +59,13 @@ def test_decode_single_jetstream_pytorch_slow(params, do_sample):
             sequence_length=512,
             expected_text="манaminationVariableßer Rog malesazine longふ Toy Champions enero Facereverse▲verbose prosecut literally disappearedअ",
         ),
+        DecodeTestParams(
+            model_id="Trendyol/Trendyol-LLM-7b-base-v0.1",
+            sequence_length=512,
+            expected_text="\nThe clocks were striking thirteen, and the clocks were striking thirteen.",
+        ),
     ],
-    ids=["TinyLLama-v0", "gemma-2b", "Mixtral-tiny"],
+    ids=["TinyLLama-v0", "gemma-2b", "Mixtral-tiny", "Trendyol-LLM-7b-base-v0.1"],
 )
 def test_decode_single_jetstream_pytorch(params, do_sample):
     params.do_sample = do_sample

